docs/api/ref/GetAgent.md

```diff
@@ -0,0 +1,125 @@
+
+```Typescript
+/** 
+ * Retrieves an FDC3 DesktopAgent instance either from a Desktop Agent that supports
+ * injection or by using the FDC3 Web Connection Protocol (WCP) to establish a connection
+ * with a browser-resident Desktop Agent.
+ * 
+ * @param {GetAgentParams} params Required parameters for the function.
+ *  
+ * @return A promise that resolves to an object which contains a DesktopAgent, or which
+ * rejects with a string from the AgentError union if a DesktopAgent cannot be established.
+ * 
+ * @example  
+ * const { desktopAgent: fdc3 } = await getAgent({ 
+ *     appId: “myApp@myorg.com/appd”
+ * }); 
+ * 
+ *  * @example Using appDUrl
+ * const { desktopAgent: fdc3 } = await getAgent({ 
+ *     appId: “myApp”,
+ *     appDUrl: "myorg.com/appd"
+ * }); 
+ */
+
+export type GetAgentFunction = ( 
+	params: GetAgentParams,  
+) => Promise<{
+    desktopAgent: DesktopAgent,
+    implementationMetadata: ImplementationMetadata,
+    appMetadata: AppMetadata
+}>;
+
+/** 
+ * @typedef {Object} GetAgentParams Type representing parameters passed to the 
+ * getAgent function.
+ * 
+ * @property {string} appId The fully qualified appId that represents the application.
+ * (in the form <app id>@<appd origin>)
+ * 
+ * @property {URL} appDUrl A URL that points to an appD record for the application.
+ * Used as an alternative to providing a fully qualified appId.
+ * 
+ * @property {number} timeout Number of milliseconds to allow for establishing a 
+ * DesktopAgent. When the timeout expires, the optional provided failover function will be
+ * run. Default 750.
+ * 
+ * @property {boolean} channelSelector Flag indicating that the application  
+ * requires getAgent() to create a channel selector UI. Defaults to true.
+ * 
+ * @property {boolean} intentResolver Flag indicating that the application  
+ * requires getAgent() to create an intent resolver UI. Defaults to true.
+ * 
+ * @property {function} failover A optional function that can establish connectivity
+ * to a DesktopAgent if standard mechanisms fail. If a WindowProxy or URL is provided
+ * then getAgent() will re-run its internal algorithm with those objects (restarting
+ * the timeout). The function may also simply return a DesktopAgent.
+ */ 
+
+export type GetAgentParams = {
+    timeout ?: number, // Defaults to 750
+    appId ?: string,
+    appDUrl ?: string,
+    failover ?: (args: GetAgentParams) =>  Promise<WindowProxy | URL | DesktopAgent>
```

Comment:
I'm not forgetting - it's just very difficult to bring these up in the WG.  Normally I have about 5 minutes to talk at the end of a meeting and I've got other things I want to put across.

Reply:
Well you know how to get an item on the agenda. If you want to discuss it let me know.

Reply:
Having reviewed your loader approach I still prefer what we're doing here and agreed in the discussion group - although the approaches are similar the difference appears to be that you are breaking out the connection strategies so that the app developer has to pass them in, where we designed this function to attempt the built-in strategies and to expose the failover to allow other approaches to be added for advanced cases (like using an adaptor to connect to a container, or to start a DA), or to prepare the app to run without a DA. That allows for a simpler usage for most apps (which won't provide a failover) or to simplify the implementation of one (e.g. just return a URL).

What you've implemented makes a lot of sense internally, but I think the optional failover function simpler for app developers to understand and implement (although theres not a huge difference).

Reply:
Also with the loaders approach its not clear what the sequencing is - we worked through racin the two basic strategies (injected FDC3 vs. a parent window/frame) as they are mutually exclusive and then running the failover afterwards allowing a 'parent' DA to take precedence over an alternative strategy (one of the first things we agreed in the discussion group as a MUST) - it appears you just race those provided (with Promise.any, which is the correct way to do that, here https://github.com/finos-labs/fdc3-for-the-web/blob/4242e9cfb6d0de43cda18af642c8977f629a46de/packages/client/src/index.ts#L33-L34) - which is exactly as planned for the built-in strategies but not the failover, which should come after... Otherwise, a rapidly returning alternative strategy would always override the defaults - and if you're going to do that you don't need the getAgent function at all...

Reply:
But like your Loader implementation for use inside the library!

Reply:
but on reflection I'll drop the ability to return void from a failover (that I was just adding in) as its not necessary - getAgent will reject its promise with a reason if no agent is found - that should be the way to handle preparing an app for running without an agent